### DISCUSSION TITLE
Translate URL to <Link> href

### DISCUSSION THREAD

**Discussion by JonCognioDigital** [vercel/next.js#14567] [Help]
Is there any way to take a fully formed URL and work out which template it should use?

I'll explain my use case:

When users put links into our CMS they obviously use the proper url, e.g. **"domain.com/section/page-name**", so when I pull this out of the CMS and attempt to make a **<Link>** tag out of it I have to do the following....

`<Link href="???" as={cmsContent.linkUrl}>`

Because any URL can come out of the CMS I somehow have to map it to the correct template, by using some kind of matching service. In this case **"domain.com/section/page-name"** should have **href="/section/[...slug]"**

It seems like creating a lookup table goes against the who point of Next's dynamic routing which doesn't require a central routing table. I'm not sure if there's a built in way to do this. Instinct tells me there should be because when a user types a URL into the browser they get the correct page, so Next must already be matching URLs to templates internally.

Any ideas?

**Comment by simplenotezy** [comment]
Looking for solutions to this too.

**Comment by simplenotezy** [comment]
I've found a workaround, although not the prettiest, it does it's job.

The main idéa is this:

- If it is a local link, e.g. a link to your own application, it will contain either: a) The host of the current browser window b) A reference to your API endpoint (`API_ENDPOINT`), in my case, all links that _should_ be local, contains the hostname from `process.env.API_ENDPOINT`, so it was rather easy to check if host matched, and if so:
- Only return the "pathname" of the URL, that way the Next.js Router will handle the navigation smoothly. I've tested providing a full absolute URL to an external site to the Router.push method, and that even seems to work as well.


	  const resolveLocalPathFromURL = (link) => {
		let href = link || '#';

		if (href && !href.startsWith('#') && !['/', '#'].includes(href)) {
			try {
				const url = new URL(href);
				if (
					url.host.includes(new URL(process.env.API_ENDPOINT).host) ||
					(typeof window !== 'undefined' &&
						window.location &&
						url.host.includes(window.location.host))
				) {
					href = url.pathname;
				}
			} catch (error) {
				// eslint-disable-next-line no-console
				console.error(
					'Error happened trying to handle dynamic link url:',
					href,
					error
				);
			}
		}

		return href;
	   };

	   // eslint-disable-next-line import/prefer-default-export
	   export { resolveLocalPathFromURL };

Then I have an `useEffect` on the component(s) that have user generated content, in my case it's only one component, so it is easy enough:

	const dynamicContentLinksEventHandler = (e, router) => {
		e.preventDefault();
		router.push(resolveLocalPathFromURL(e.target.href));
	};

	const router = useRouter();

	useEffect(() => {
		if (value.includes('<a') && element && element.current) { // replace "value" with the prop that (may) contain links
			const linkElements = element.current.getElementsByTagName('a');
			for (const link of linkElements) {
				if (!link.classList.contains('ex-h')) {
				// link.removeEventListener('click', dynamicContentLinksEventHandler) // does not work for some reason
				link.addEventListener('click', (e) =>
					dynamicContentLinksEventHandler(e, router)
				);
				link.classList.add('ex-h'); // add flag, to prevent double event listeners. for some reason removeEventListener does not work
				}
			}
		}
	}, [value, router]);
     [...]
     <div ref={element} dangerouslySetInnerHTML={{ __html: value } } />
